### PR TITLE
feat(event_handlers): Add support for Lambda Function URLs

### PR DESCRIPTION
**Pull request by rubenfonseca**
**Issue number:** #1142, #1141

### Changes

> Please provide a summary of what's being changed

This PR adds support for [Lambda Function URL](https://docs.aws.amazon.com/lambda/latest/dg/urls-configuration.html) events (data class) and event handlers.

#### Data Class

carbon (7)

For now, it seems that Lambda Function URL events [follow the exact same payload as HTTP APIs Payload Format Version 2.0](https://docs.aws.amazon.com/lambda/latest/dg/urls-invocation.html#urls-payloads).

#### Event Handler

carbon (5)
carbon (6)

Certain keys in this payload format don't make sense for function urls (e.g: `routeKey`, `stage`). These keys will have default values that come on the payload, but they are not useful since they can't be changed.

### User experience

> Please share what the user experience looks like before and after this change

Users could already use Lambda Function URLs before by using the API Gateway HTTP API data classes and handlers. However, we make the usage explicit and prepare for any future deviation from the API Gateway format.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

-----
[View rendered docs/core/event_handler/api_gateway.md](https://github.com/rubenfonseca/aws-lambda-powertools-python/blob/feat/lambda-function-urls/docs/core/event_handler/api_gateway.md)
[View rendered docs/index.md](https://github.com/rubenfonseca/aws-lambda-powertools-python/blob/feat/lambda-function-urls/docs/index.md)
[View rendered docs/utilities/data_classes.md](https://github.com/rubenfonseca/aws-lambda-powertools-python/blob/feat/lambda-function-urls/docs/utilities/data_classes.md)

### CODE REVIEW

**Comment by gwlester** [comment]
Looks good -- thanks much!!!!

File: aws_lambda_powertools/event_handler/lambda_function_url.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    - https://docs.aws.amazon.com/lambda/latest/dg/urls-configuration.html
+    - https://docs.aws.amazon.com/lambda/latest/dg/urls-invocation.html#urls-payloads
+
+    Examples
+    --------
+    Simple example with a custom lambda handler using the Tracer capture_lambda_handler decorator
+
+    ```python
+    from aws_lambda_powertools import Tracer
+    from aws_lambda_powertools.event_handler import LambdaFunctoinUrlResolver
```
Comment (jplock) [inline aws_lambda_powertools/event_handler/lambda_function_url.py]: spelling: `LambdaFunctoinUrlResolver` -> `LambdaFunctionUrlResolver`

File: aws_lambda_powertools/event_handler/lambda_function_url.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    tracer = Tracer()
+    app = LambdaFunctoinUrlResolver()
```
Comment (jplock) [inline aws_lambda_powertools/event_handler/lambda_function_url.py]: spelling: `LambdaFunctoinUrlResolver` -> `LambdaFunctionUrlResolver`

**Review by jplock — COMMENTED** [review]
Please correct spelling errors

File: aws_lambda_powertools/event_handler/api_gateway.py [hunk 3]
```diff
@@ -546,6 +552,9 @@ def _to_proxy_event(self, event: Dict) -> BaseProxyEvent:
         if self._proxy_type == ProxyEventType.APIGatewayProxyEventV2:
             logger.debug("Converting event to API Gateway HTTP API contract")
             return APIGatewayProxyEventV2(event)
+        if self._proxy_type == ProxyEventType.LambdaFunctionUrlEvent:
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/api_gateway.py:555]: note to future self: we should improve the return type to a more accurate one with `@overload`.

File: aws_lambda_powertools/event_handler/lambda_function_url.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+from aws_lambda_powertools.event_handler.api_gateway import ApiGatewayResolver, ProxyEventType
+from aws_lambda_powertools.utilities.data_classes import LambdaFunctionUrlEvent
+
+
+class LambdaFunctionUrlResolver(ApiGatewayResolver):
+    """AWS Lambda Function URL resolver
+
+    Notes:
+    -----
+    For now, this seems to work the same way as API Gateway HTTP APIs Payload Format Version 2.0.
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/lambda_function_url.py]:
```suggestion
    Lambda Function URL follows the API Gateway HTTP APIs Payload Format Version 2.0.
```

File: aws_lambda_powertools/event_handler/lambda_function_url.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    Documentation:
+    - https://docs.aws.amazon.com/lambda/latest/dg/urls-configuration.html
+    - https://docs.aws.amazon.com/lambda/latest/dg/urls-invocation.html#urls-payloads
+
+    Examples
+    --------
+    Simple example with a custom lambda handler using the Tracer capture_lambda_handler decorator
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/lambda_function_url.py]:
```suggestion
    Simple example integrating with Tracer
```

File: aws_lambda_powertools/logging/correlation_paths.py [hunk 6]
```diff
@@ -6,4 +6,5 @@
 APPSYNC_RESOLVER = 'request.headers."x-amzn-trace-id"'
 APPLICATION_LOAD_BALANCER = 'headers."x-amzn-trace-id"'
 EVENT_BRIDGE = "id"
+LAMBDA_FUNCTION_URL = "requestContext.requestId"
```
Comment (heitorlessa) [inline aws_lambda_powertools/logging/correlation_paths.py]:
```suggestion
LAMBDA_FUNCTION_URL = API_GATEWAY_REST
```

As they follow the same payload specification

File: aws_lambda_powertools/event_handler/lambda_function_url.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    app = LambdaFunctionUrlResolver()
+
+    @app.get("/get-call")
+    def simple_get():
+        return {"message": "Foo"}
+
+    @app.post("/post-call")
+    def simple_post():
+        post_data: dict = app.current_event.json_body
+        return {"message": post_data["value"]}
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/lambda_function_url.py]:
```suggestion
        return {"message": post_data}
```

File: aws_lambda_powertools/utilities/data_classes/api_gateway_proxy_event.py [hunk 8]
```diff
@@ -127,19 +127,22 @@ def caller_id(self) -> Optional[str]:
     def cognito_amr(self) -> Optional[List[str]]:
         """This represents how the user was authenticated.
         AMR stands for  Authentication Methods References as per the openid spec"""
-        return self["cognitoIdentity"].get("amr")
+        cognito_identity = self["cognitoIdentity"] or {}
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/api_gateway_proxy_event.py]:
```suggestion
        cognito_identity = self["cognitoIdentity"] or {}  # not available in FunctionURL
```

File: aws_lambda_powertools/utilities/data_classes/api_gateway_proxy_event.py [hunk 9]
```diff
@@ -127,19 +127,22 @@ def caller_id(self) -> Optional[str]:
...
+        return cognito_identity.get("amr")
 
     @property
     def cognito_identity_id(self) -> Optional[str]:
         """The Amazon Cognito identity ID of the caller making the request.
         Available only if the request was signed with Amazon Cognito credentials."""
-        return self["cognitoIdentity"].get("identityId")
+        cognito_identity = self.get("cognitoIdentity") or {}
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/api_gateway_proxy_event.py]:
```suggestion
        cognito_identity = self.get("cognitoIdentity") or {}  # not available in FunctionURL
```

File: aws_lambda_powertools/utilities/data_classes/api_gateway_proxy_event.py [hunk 10]
```diff
@@ -127,19 +127,22 @@ def caller_id(self) -> Optional[str]:
...
+        return cognito_identity.get("identityId")
 
     @property
     def cognito_identity_pool_id(self) -> Optional[str]:
         """The Amazon Cognito identity pool ID of the caller making the request.
         Available only if the request was signed with Amazon Cognito credentials."""
-        return self["cognitoIdentity"].get("identityPoolId")
+        cognito_identity = self.get("cognitoIdentity") or {}
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/api_gateway_proxy_event.py]:
```suggestion
        cognito_identity = self.get("cognitoIdentity") or {}  # not available in FunctionURL
```

File: aws_lambda_powertools/utilities/data_classes/api_gateway_proxy_event.py [hunk 11]
```diff
@@ -159,12 +162,14 @@ def user_id(self) -> Optional[str]:
 
 class RequestContextV2Authorizer(DictWrapper):
     @property
-    def jwt_claim(self) -> Dict[str, Any]:
-        return self["jwt"]["claims"]
+    def jwt_claim(self) -> Optional[Dict[str, Any]]:
+        jwt = self.get("jwt") or {}
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/api_gateway_proxy_event.py]:
```suggestion
        jwt = self.get("jwt") or {}  # not available in FunctionURL
```

File: aws_lambda_powertools/utilities/data_classes/api_gateway_proxy_event.py [hunk 12]
```diff
@@ -159,12 +162,14 @@ def user_id(self) -> Optional[str]:
...
+        return jwt.get("claims")
 
     @property
-    def jwt_scopes(self) -> List[str]:
-        return self["jwt"]["scopes"]
+    def jwt_scopes(self) -> Optional[List[str]]:
+        jwt = self.get("jwt") or {}
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/api_gateway_proxy_event.py]:
```suggestion
        jwt = self.get("jwt") or {}  # not available in FunctionURL
```

File: aws_lambda_powertools/utilities/data_classes/common.py [hunk 13]
```diff
@@ -395,5 +395,6 @@ def time_epoch(self) -> int:
     @property
     def authentication(self) -> Optional[RequestContextClientCert]:
         """Optional when using mutual TLS authentication"""
-        client_cert = self["requestContext"].get("authentication", {}).get("clientCert")
+        authentication = self["requestContext"].get("authentication") or {}
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/common.py:399]:
```suggestion
        # FunctionURL might have NONE as AuthZ
        authentication = self["requestContext"].get("authentication") or {}
```

File: aws_lambda_powertools/utilities/data_classes/lambda_function_url_event.py [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class LambdaFunctionUrlEvent(APIGatewayProxyEventV2):
+    """AWS Lambda Function URL event
+
+    Notes:
+    -----
+    For now, this seems to follow the exact same payload as HTTP APIs Payload Format Version 2.0.
+    Certain keys in this payload format don't make sense for function urls (e.g: `routeKey`, `stage`).
+    These keys will have default values that come on the payload, but they are not useful since they can't be changed.
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/lambda_function_url_event.py]:
```suggestion
    Lambda Function URL follows the API Gateway HTTP APIs Payload Format Version 2.0.
    
    Keys related to API Gateway features not available in Function URL use a sentinel value (e.g.`routeKey`, `stage`).
```

Tip: Write first, edit to be more assertive (is or isnt?), and then settle for the shortest with the same meaning.

File: docs/core/event_handler/api_gateway.md [hunk 15]
```diff
@@ -87,6 +95,22 @@ When using Amazon Application Load Balancer (ALB) to front your Lambda functions
 --8<-- "examples/event_handler_rest/src/getting_started_alb_api_resolver.py"
 ```
 
+#### Lambda Function URL
+
+When using an [AWS Lambda Function URL](https://docs.aws.amazon.com/lambda/latest/dg/urls-configuration.html), you can use `LambdaFunctionUrlResolver`.
```
Comment (heitorlessa) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
When using [AWS Lambda Function URL](https://docs.aws.amazon.com/lambda/latest/dg/urls-configuration.html), you can use `LambdaFunctionUrlResolver`.
```

File: docs/core/event_handler/api_gateway.md [hunk 16]
```diff
@@ -270,7 +294,7 @@ This will ensure that CORS headers are always returned as part of the response w
 
 #### Pre-flight
 
-Pre-flight (OPTIONS) calls are typically handled at the API Gateway level as per [our sample infrastructure](#required-resources), no Lambda integration necessary. However, ALB expects you to handle pre-flight requests.
+Pre-flight (OPTIONS) calls are typically handled at the API Gateway or Lambda Function URL level as per [our sample infrastructure](#required-resources), no Lambda integration necessary. However, ALB expects you to handle pre-flight requests.
```
Comment (heitorlessa) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
Pre-flight (OPTIONS) calls are typically handled at the API Gateway or Lambda Function URL level as per [our sample infrastructure](#required-resources), no Lambda integration is necessary. However, ALB expects you to handle pre-flight requests.
```

Fixing a old typo :D

File: docs/core/event_handler/api_gateway.md [hunk 17]
```diff
@@ -338,6 +362,7 @@ Like `compress` feature, the client must send the `Accept` header with the corre
 
 ???+ warning
     This feature requires API Gateway to configure binary media types, see [our sample infrastructure](#required-resources) for reference.
+    For Lambda Function URLs, no additional configuration is necessary.
```
Comment (heitorlessa) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
???+ warning
    This feature requires API Gateway to configure binary media types, see [our sample infrastructure](#required-resources) for reference.

???+ note
    Lambda Function URLs handle binary media types automatically.
```

In this case, it's best so separate what's a warning vs a note - "Do something vs Business as usual"

File: docs/core/event_handler/api_gateway.md [hunk 18]
```diff
@@ -380,7 +405,7 @@ This will enable full tracebacks errors in the response, print request and respo
 
 ### Custom serializer
 
-You can instruct API Gateway handler to use a custom serializer to best suit your needs, for example take into account Enums when serializing.
+You can instruct an event handler to use a custom serializer to best suit your needs, for example take into account Enums when serializing.
```
Comment (heitorlessa) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
You can instruct event handler to use a custom serializer to best suit your needs, for example take into account Enums when serializing.
```

Great catch!!!!

File: docs/core/event_handler/api_gateway.md [hunk 19]
```diff
@@ -501,7 +526,7 @@ A micro function means that your final code artifact will be different to each f
 
 **Downsides**
 
-* **Upfront investment**. You need custom build tooling to bundle assets, including [C bindings for runtime compatibility](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html){target="_blank"}. `Operations become more elaborate — you need to standardize tracing labels/annotations, structured logging, and metrics to pinpoint root causes.
+* **Upfront investment**. You need custom build tooling to bundle assets, including [C bindings for runtime compatibility](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html){target="_blank"}. Operations become more elaborate — you need to standardize tracing labels/annotations, structured logging, and metrics to pinpoint root causes.
```
Comment (heitorlessa) [inline docs/core/event_handler/api_gateway.md:530]: great catch!!

File: docs/index.md [hunk 20]
```diff
@@ -411,22 +411,22 @@ sam init --location https://github.com/aws-samples/cookiecutter-aws-sam-python
...
+| [Metrics](./core/metrics.md)                                                                                                                             | Custom Metrics created asynchronously via CloudWatch Embedded Metric Format (EMF)                                                                         |
+| [Event handler: AppSync](./core/event_handler/appsync.md)                                                                                                | AppSync event handler for Lambda Direct Resolver and Amplify GraphQL Transformer function                                                                 |
+| [Event handler: API Gateway, ALB and Lambda Function URL](https://awslabs.github.io/aws-lambda-powertools-python/latest/core/event_handler/api_gateway/) | Amazon API Gateway REST/HTTP API and ALB event handler for Lambda functions invoked using Proxy integration                                               |
```
Comment (heitorlessa) [inline docs/index.md]:
```suggestion
| [Event handler: API Gateway, ALB and Lambda Function URL](https://awslabs.github.io/aws-lambda-powertools-python/latest/core/event_handler/api_gateway/) | Amazon API Gateway REST/HTTP API and ALB event handler for Lambda functions invoked using Proxy integration, and Lambda Function URL                                               |
```

File: examples/event_handler_lambda_function_url/sam/template.yaml [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+    Tracing: Active
+    Environment:
+      Variables:
+        LOG_LEVEL: INFO
+        POWERTOOLS_LOGGER_SAMPLE_RATE: 0.1
+        POWERTOOLS_LOGGER_LOG_EVENT: true
+        POWERTOOLS_SERVICE_NAME: example
+    FunctionUrlConfig:
+      Cors: # see CORS section
+        # Notice that values here are Lists of Strings, vs comma-separated values on API Gateway
```
Comment (heitorlessa) [inline examples/event_handler_lambda_function_url/sam/template.yaml:18]: Great note!

File: examples/event_handler_lambda_function_url/sam/template.yaml [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+
+Resources:
+  ApiFunction:
+    Type: AWS::Serverless::Function
+    Properties:
+      Handler: getting_started_lambda_function_url_resolver.lambda_handler
+      CodeUri: ../src
+      Description: API handler function
+      FunctionUrlConfig:
+        AuthType: NONE
```
Comment (heitorlessa) [inline examples/event_handler_lambda_function_url/sam/template.yaml]:
```suggestion
        AuthType: NONE  # AWS_IAM for added security beyond sample documentation
```

**Review by heitorlessa — CHANGES_REQUESTED** [review]
looks great, thank you @rubenfonseca !!!! Minor suggestions that you can easily commit, then we're good to merge and release tomorrow!

File: aws_lambda_powertools/event_handler/api_gateway.py (hunk 3 above)
Comment (rubenfonseca) [reply aws_lambda_powertools/event_handler/api_gateway.py:555]: note to future self: wait for python 3.10 to actually have [proper pattern matching](https://peps.python.org/pep-0634/)

**Comment by heitorlessa** [comment]
YES except we won’t be able to use it (unless someone backported it? :/

**Review by heitorlessa — APPROVED** [review]

**Comment by heitorlessa** [comment]
Merging! Should be out in our next release tomorrow

**Pull request merged**